pandas/core/reshape/reshape.py

```diff
@@ -743,7 +746,16 @@ def _convert_level_number(level_num: int, columns: Index):
         if slice_len != levsize:
             chunk = this.loc[:, this.columns[loc]]
             chunk.columns = level_vals_nan.take(chunk.columns.codes[-1])
-            value_slice = chunk.reindex(columns=level_vals_used).values
+            # Override fill value to prevent upcasting to float64
+            # if we have lower precision floats
+            common_type = find_common_type(chunk.dtypes.tolist())
+            if common_type.kind == "f" and not is_extension_array_dtype(common_type):
+                fill_value = common_type.type(np.nan)
+            else:
+                fill_value = np.nan
```

Comment:
Any reason for not using common_dtype.na_value?

Right now "Float32" would add np.nan, but this should be pd.NA, Same for Int64 for example

Reply:
The default fill_value is np.nan for reindex, even for EAs. That's probably something to fix elsewhere, though (IIRC, the fill value gets converted somewhere in the EA methods, maybe take? It's been a while since I looked into it though).

Reply:
Updated.